src/rpt3_euler_mapgrid.f90

```diff
@@ -0,0 +1,295 @@
+  !
+  ! Compute the Roe-averaged variables needed in the Roe solver.
+
+  ! Loop over grid cell interfaces
+  DO i = 2-mbc, mx+mbc
+     IF (qr(1,i-1) <= 0.d0 .OR. ql(1,i) <= 0.d0) THEN
+        WRITE(*,*) i, mu, mv, mw
+        WRITE(*,'(5e12.4)') (qr(j,i-1),j=1,5)
+        WRITE(*,'(5e12.4)') (ql(j,i),j=1,5)
+        IF (ixyz == 1) WRITE(6,*) '*** (rpt3) rho <= 0 in x-sweep at ',i
+        IF (ixyz == 2) WRITE(6,*) '*** (rpt3) rho <= 0 in y-sweep at ',i
+        IF (ixyz == 3) WRITE(6,*) '*** (rpt3) rho <= 0 in z-sweep at ',i
+        WRITE(6,*) 'stopped with rho <= 0...'
+        STOP
+     ENDIF
+     rhsqrtl = SQRT(qr(1,i-1))
+     rhsqrtr = SQRT(ql(1,i))
+     pl = gamma1*(qr(5,i-1) - 0.5d0*(qr(mu,i-1)**2 + &
+          qr(mv,i-1)**2 + qr(mw,i-1)**2)/qr(1,i-1))
+     pr = gamma1*(ql(5,i) - 0.5d0*(ql(mu,i)**2 + &
+          ql(mv,i)**2 + ql(mw,i)**2)/ql(1,i))
+     rhsq2 = rhsqrtl + rhsqrtr
+     u(i) = (qr(mu,i-1)/rhsqrtl + ql(mu,i)/rhsqrtr) / rhsq2
+     v(i) = (qr(mv,i-1)/rhsqrtl + ql(mv,i)/rhsqrtr) / rhsq2
+     w(i) = (qr(mw,i-1)/rhsqrtl + ql(mw,i)/rhsqrtr) / rhsq2
+     enth(i) = (((qr(5,i-1)+pl)/rhsqrtl &
+          + (ql(5,i)+pr)/rhsqrtr)) / rhsq2
+     u2v2w2(i) = u(i)**2 + v(i)**2 + w(i)**2
+     asqrd = gamma1*(enth(i) - .5d0*u2v2w2(i))
+     
+     IF (i>=0 .AND. i<=mx .AND. asqrd <= 0.d0) THEN
+        IF (ixyz == 1) WRITE(6,*) '*** (rpt3) a**2 <= 0 in x-sweep at ',i
+        IF (ixyz == 2) WRITE(6,*) '*** (rpt3) a**2 <= 0 in y-sweep at ',i
+        IF (ixyz == 3) WRITE(6,*) '*** (rpt3) a**2 <= 0 in z-sweep at ',i
+        WRITE(6,*) 'stopped with a**2 < 0...'
+        STOP
+     ENDIF
+     a(i) = SQRT(asqrd)
+     g1a2(i) = gamma1 / asqrd
+     euv(i) = enth(i) - u2v2w2(i)
+  END DO
+
+  ! Solve the Riemann Problem
+  DO i = 2-mbc, mx+mbc
+
+     nx = aux2(nu,i,2)
+     ny = aux2(nv,i,2)
+     nz = aux2(nw,i,2)
+     vn = u(i)*nx + v(i)*ny + w(i)*nz
+     IF(icoor == 2)THEN
+        areaRatioMinus = aux2(iratio,i-2+imp,2)
+        areaRatioPlus  = aux3(iratio,i-2+imp,2)
+     ELSEIF(icoor == 3)THEN
+        areaRatioMinus = aux2(iratio,i-2+imp,2)
+        areaRatioPlus  = aux2(iratio,i-2+imp,3)
+     ENDIF
+     ! Pick the largest normal to avoid using a singular direction
+     IF(ABS(nx) >= ABS(ny) .AND. ABS(nx) >= ABS(nz))THEN
+        nonsingularNormal = "x"
+     ELSEIF(ABS(ny) >= ABS(nx) .AND. ABS(ny) >= ABS(nz))THEN
+        nonsingularNormal = "y"
+     ELSEIF(ABS(nz) >= ABS(nx) .AND. ABS(nz) >= ABS(ny))THEN
+        nonsingularNormal = "z"
+     ELSE
+        WRITE(*,*)"Invalid Normal Vector..."
+        WRITE(*,*)"nx: ",nx,"ny: ",ny,"nz: ",nz
+        STOP
+     END IF
+     
+     beta(4) = g1a2(i)*(euv(i)*asdq(1,i) &
+          + u(i)*asdq(mu,i) + v(i)*asdq(mv,i) + w(i)*asdq(mw,i) - asdq(5,i))
+     beta(5) = ( (a(i)-vn)*asdq(1,i) - a(i)*beta(4) &
+          + nx*asdq(mu,i) + ny*asdq(mv,i) + nz*asdq(mw,i) )/(2.d0*a(i))
+     beta(1) = asdq(1,i) - beta(4) - beta(5)
+     
+     SELECT CASE(TRIM(nonsingularNormal))
```

Comment:
This is minimal as well but doing a string comparison this deep into the code can lead to significant slow-downs.  I have found the best option to maintain flexibility and readability is to use function pointers for these type of situations.  Check out how https://github.com/clawpack/geoclaw/blob/master/src/2d/shallow/surge/storm_module.f90 uses function pointers to do something similar.